File: java/breadth-first-search/FrogJump.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Solution {
    public boolean canCross(int[] stones) {
        int destination = stones[stones.length - 1];
        
        Map<Integer, Set<Integer>> hashMap = new HashMap<>();
        for(int stone : stones) {
            hashMap.put(stone, new HashSet<>());
        }
        
        hashMap.get(stones[0]).add(1);
        
        for(int i = 0; i < stones.length; i++) {
            int currStonePosition = stones[i];
            for(int jump : hashMap.get(currStonePosition)) {
                int nextPosition = currStonePosition + jump;
                if(nextPosition == destination) {
                    return true;
                }
                if(hashMap.containsKey(nextPosition)) {
                    if(jump - 1 > 0) {
                        hashMap.get(nextPosition).add(jump - 1);
                    }
                    hashMap.get(nextPosition).add(jump);
                    hashMap.get(nextPosition).add(jump + 1);
                }   
            }
        }

        return false;
    }
}

File: python/breadth-first-search/FrogJump.py
from typing import List

class Solution:
    def canCross(self, stones: List[int]) -> bool:

        destination = stones[-1]

        hash_map = dict()
        for stone in stones:
            hash_map[stone] = set()
        
        hash_map[stones[0]].add(1)

        for i in range(len(stones)):
            curr_stone_position = stones[i]
            for jump in hash_map[curr_stone_position]:
                next_position = curr_stone_position + jump
                if next_position == destination:
                    return True
                if next_position in hash_map:
                    if jump - 1 > 0:
                        hash_map[next_position].add(jump - 1)
                    hash_map[next_position].add(jump)
                    hash_map[next_position].add(jump + 1)

        return False

        